kentik_api_library/kentik_api/utils/device_cache.py

```diff
@@ -0,0 +1,140 @@
+import logging
+import pickle
+import sys
+from pathlib import Path
+from typing import Dict, Generator, List, Optional, Tuple, Type, TypeVar
+
+from kentik_api import KentikAPI
+from kentik_api.public import Device, DeviceInterface
+
+log = logging.getLogger("device_cache")
+
+
+class DeviceCacheIterator:
+    def __init__(self, devices) -> None:
+        self.devices = devices
+        self.id_list = sorted([d.id for d in devices.all])
+
+    def __next__(self):
+        try:
+            return self.devices[self.id_list.pop(0)]
+        except IndexError:
+            raise StopIteration
+
+
+T = TypeVar("T", bound="DeviceCache")
+
+
+class DeviceCache:
+    @classmethod
+    def from_api(cls: Type[T], api: KentikAPI, labels: Optional[List[str]] = None) -> T:
+        log.debug("Fetching all devices")
+        devices = api.devices.get_all()
+        return cls(devices, labels)
+
+    @classmethod
+    def from_pickle(cls: Type[T], filename: str) -> T:
+        file = Path(filename)
+        log.debug(f"Reading device data from {file.resolve()}")
+        return pickle.load(file.open("rb"))
```

Comment:
Shouldn't close the file afterwards?

Reply:
Based on my quick check, the filedescriptor is closed as soon as it goes out of scope, but I'll rather add context manager to be completely safe.